### PR TITLE
[SystemZ][z/OS] Add new openFileForReadBinary function, and pass IsText parameter to getBufferForFile

### PR DESCRIPTION
**Pull request by abhina-sree**
This patch adds an IsText parameter to the following getBufferForFile, getBufferForFileImpl. We introduce a new virtual function openFileForReadBinary which defaults to openFileForRead except in RealFileSystem which uses the OF_None flag instead of OF_Text.

The default is set to OF_Text instead of OF_None, this change in value does not affect any other platforms other than z/OS. Setting this parameter correctly is required to open files on z/OS in the correct encoding. The IsText parameter is based on the context of where we open files, for example, in the ASTReader, HeaderMap requires that files always be opened in binary even though they might be tagged as text.

### CODE REVIEW

File: llvm/lib/Support/VirtualFileSystem.cpp [hunk 1]
```diff
@@ -117,8 +117,12 @@ FileSystem::~FileSystem() = default;
 
 ErrorOr<std::unique_ptr<MemoryBuffer>>
 FileSystem::getBufferForFile(const llvm::Twine &Name, int64_t FileSize,
-                             bool RequiresNullTerminator, bool IsVolatile) {
-  auto F = openFileForRead(Name);
+                             bool RequiresNullTerminator, bool IsVolatile,
+                             bool IsText) {
+  auto openFileFunctionPointer = &FileSystem::openFileForRead;
+  if (!IsText)
+    openFileFunctionPointer = &FileSystem::openFileForReadBinary;
```
Comment (perry-ca) [inline llvm/lib/Support/VirtualFileSystem.cpp:124]:
```suggestion
  auto F = IsText ?openFileForRead(Name) : openFileForReadBinary(Name);
```

File: llvm/lib/Support/VirtualFileSystem.cpp [hunk 2]
```diff
@@ -324,6 +330,17 @@ ErrorOr<Status> RealFileSystem::status(const Twine &Path) {
 
 ErrorOr<std::unique_ptr<File>>
 RealFileSystem::openFileForRead(const Twine &Name) {
+  SmallString<256> RealName, Storage;
```
Comment (perry-ca) [inline llvm/lib/Support/VirtualFileSystem.cpp:333]: Can you put this code into a static function or even a private member function so you can avoid the duplication of code.

File: llvm/lib/Support/VirtualFileSystem.cpp (hunk 1 above)
Comment (abhina-sree) [reply llvm/lib/Support/VirtualFileSystem.cpp]: Thanks this is done

File: llvm/lib/Support/VirtualFileSystem.cpp (hunk 2 above)
Comment (abhina-sree) [reply llvm/lib/Support/VirtualFileSystem.cpp]: I've created a private function for both functions to call, thanks!

File: llvm/lib/Support/VirtualFileSystem.cpp [hunk 3]
```diff
@@ -302,6 +305,17 @@ class RealFileSystem : public FileSystem {
     return Storage;
   }
 
+  ErrorOr<std::unique_ptr<File>> openFileForRead(const Twine &Name,
```
Comment (perry-ca) [inline llvm/lib/Support/VirtualFileSystem.cpp:308]: Thanks, Could you mark this as static.

Comment (abhina-sree) [reply llvm/lib/Support/VirtualFileSystem.cpp:308]: I don't think I'm able to mark it static because it calls adjustPath which is non-static

Comment (perry-ca) [reply llvm/lib/Support/VirtualFileSystem.cpp:308]: np.  I was misreading the diff.  I didn't see it was a private member function.  It looked like a file scope function.  I was thinking file scope static, not static member.

Comment (kadircet) [inline llvm/lib/Support/VirtualFileSystem.cpp:308]: it was quite confusing to see this as an overload, can you call it `openFileForReadWithFlags` ?

File: llvm/include/llvm/Support/VirtualFileSystem.h [hunk 4]
```diff
@@ -271,15 +271,24 @@ class FileSystem : public llvm::ThreadSafeRefCountedBase<FileSystem>,
...
   virtual llvm::ErrorOr<Status> status(const Twine &Path) = 0;
 
-  /// Get a \p File object for the file at \p Path, if one exists.
+  /// Get a \p File object for the text file at \p Path, if one exists.
   virtual llvm::ErrorOr<std::unique_ptr<File>>
   openFileForRead(const Twine &Path) = 0;
 
+  /// Get a \p File objct for the binary file at \p Path, if one exists.
+  /// This function should be called instead of openFileForRead if the file
+  /// should be opened as a binary file.
```
Comment (kadircet) [inline llvm/include/llvm/Support/VirtualFileSystem.h:280]: can you rather explain the motiviating use case? e.g. "Some filesystems can perform modifications to bytes on-disk when reading as test. Reading as binary ensures contents are received as-is in such cases, on most file systems this is same as `openFileForRead`".

File: llvm/include/llvm/Support/VirtualFileSystem.h [hunk 5]
```diff
@@ -271,15 +271,24 @@ class FileSystem : public llvm::ThreadSafeRefCountedBase<FileSystem>,
...
+    return openFileForRead(Path);
+  }
+
   /// This is a convenience method that opens a file, gets its content and then
   /// closes the file.
   llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>>
   getBufferForFile(const Twine &Name, int64_t FileSize = -1,
-                   bool RequiresNullTerminator = true, bool IsVolatile = false);
+                   bool RequiresNullTerminator = true, bool IsVolatile = false,
+                   bool IsText = true);
```
Comment (kadircet) [inline llvm/include/llvm/Support/VirtualFileSystem.h:291]: can you also add some comments about what `IsText` parameter does?

File: clang/lib/Lex/HeaderMap.cpp [hunk 6]
```diff
@@ -54,7 +54,9 @@ std::unique_ptr<HeaderMap> HeaderMap::Create(FileEntryRef FE, FileManager &FM) {
   unsigned FileSize = FE.getSize();
   if (FileSize <= sizeof(HMapHeader)) return nullptr;
 
-  auto FileBuffer = FM.getBufferForFile(FE);
+  auto FileBuffer =
+      FM.getBufferForFile(FE, /*IsVolatile=*/false,
+                          /*RequiresNullTerminator=*/true, /*IsText=*/false);
```
Comment (kadircet) [inline clang/lib/Lex/HeaderMap.cpp:59]: i think it'd be great to have a test case for this that actually fails (even if only in zOS).

File: clang/include/clang/Basic/FileManager.h [hunk 7]
```diff
@@ -292,21 +292,21 @@ class FileManager : public RefCountedBase<FileManager> {
   /// MemoryBuffer if successful, otherwise returning null.
   llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>>
   getBufferForFile(FileEntryRef Entry, bool isVolatile = false,
-                   bool RequiresNullTerminator = true,
+                   bool RequiresNullTerminator = true, bool IsText = true,
```
Comment (kadircet) [inline clang/include/clang/Basic/FileManager.h:295]: can you add some comments explaining what `IsText` is about and when it should be set to `false`?

also can you make it the last parameter ?

**Review by kadircet — CHANGES_REQUESTED** [review]
thanks! I think this LG at a high-level, added some comments about the details.

I think it'd be worthwhile to add some tests (even if not now, probably going forward). as otherwise such rare use cases might regress quite frequently.

**Comment by cor3ntin** [comment]
@abhina-sree do you plan to modify embed in this PR?

File: clang/lib/Lex/HeaderMap.cpp (hunk 6 above)
Comment (abhina-sree) [reply clang/lib/Lex/HeaderMap.cpp:59]: So the following tests were previously failing on z/OS due to the binary/text issue and will be fixed by this change
clang/test/Preprocessor/headermap-rel.c
clang/test/Preprocessor/headermap-rel2.c

File: llvm/include/llvm/Support/VirtualFileSystem.h (hunk 4 above)
Comment (abhina-sree) [reply llvm/include/llvm/Support/VirtualFileSystem.h:280]: Thanks, I've updated the comments, please let me know if they need further improvement.

File: llvm/include/llvm/Support/VirtualFileSystem.h (hunk 5 above)
Comment (abhina-sree) [reply llvm/include/llvm/Support/VirtualFileSystem.h:295]: I've added a sentence, please let me know if it needs more explanation, thanks!

File: llvm/lib/Support/VirtualFileSystem.cpp (hunk 3 above)
Comment (abhina-sree) [reply llvm/lib/Support/VirtualFileSystem.cpp]: Thanks, this is done

File: clang/include/clang/Basic/FileManager.h (hunk 7 above)
Comment (abhina-sree) [reply clang/include/clang/Basic/FileManager.h]: I've added a brief comment here as well, please let me know if its sufficient or needs more, thanks!

**Comment by abhina-sree** [comment]
> @abhina-sree do you plan to modify embed in this PR?

The related lit tests all seem to be passing on z/OS right now. I'm guessing it might be because we get the contents as binary here https://github.com/llvm/llvm-project/blob/main/clang/lib/Lex/PPDirectives.cpp#L3990. I'll have to do a bit more investigation, but it seems to be working without modifications so I don't think any changes are needed. If there are changes required, I will make them in a separate PR

**Comment by abhina-sree** [comment]
> @abhina-sree do you plan to modify embed in this PR?

Actually Sean has pointed out that if we pass in non-ascii inputs, the tests do in fact fail. So I will look into passing the IsText parameter in the necessary functions. Thanks for pointing this out!

**Review by kadircet — APPROVED** [review]
thanks for bearing with me!

**Comment by abhina-sree** [comment]
@cor3ntin, I've added support for #embed handling now, I've added a z/OS only testcase clang/test/Preprocessor/embed_zos.c which is a copy of clang/test/Preprocessor/embed_art.c except it changes the file tag to be non-ascii which is now working with the latest commit.

**Pull request merged**